// @bagsfm/bags-sdk/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BagsSDK = void 0;
const token_launch_1 = require("./services/token-launch");
const bags_client_1 = require("./api/bags-client");
const state_1 = require("./services/state");
const config_1 = require("./services/config");
const fees_1 = require("./services/fees");
const partner_1 = require("./services/partner");
const trade_1 = require("./services/trade");
const solana_1 = require("./services/solana");
const fee_share_admin_1 = require("./services/fee-share-admin");
const dexscreener_1 = require("./services/dexscreener");
const incorporation_1 = require("./services/incorporation");
const auth_1 = require("./services/auth");
class BagsSDK {
    constructor(apiKey, connection, commitment = 'processed') {
        this.bagsApiClient = new bags_client_1.BagsApiClient(apiKey);
        this.tokenLaunch = new token_launch_1.TokenLaunchService(apiKey, connection, commitment);
        this.state = new state_1.StateService(apiKey, connection, commitment);
        this.config = new config_1.ConfigService(apiKey, connection, commitment);
        this.fee = new fees_1.FeesService(apiKey, connection, commitment);
        this.partner = new partner_1.PartnerService(apiKey, connection, commitment);
        this.trade = new trade_1.TradeService(apiKey, connection, commitment);
        this.solana = new solana_1.SolanaService(apiKey, connection, commitment);
        this.feeShareAdmin = new fee_share_admin_1.FeeShareAdminService(apiKey, connection, commitment);
        this.dexscreener = new dexscreener_1.DexscreenerService(apiKey, connection, commitment);
        this.incorporation = new incorporation_1.IncorporationService(apiKey, connection, commitment);
        this.auth = new auth_1.AuthService(apiKey, connection, commitment);
    }
}
exports.BagsSDK = BagsSDK;
