// @bagsfm/bags-sdk/dist/services/fees.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeesService = void 0;
const web3_js_1 = require("@solana/web3.js");
const base_1 = require("./base");
const bs58_1 = __importDefault(require("bs58"));
class FeesService extends base_1.BaseService {
    constructor(apiKey, connection, commitment = 'processed') {
        super(apiKey, connection, commitment);
    }
    async getAllClaimablePositions(wallet) {
        const response = await this.bagsApiClient.get('/token-launch/claimable-positions', {
            params: {
                wallet: wallet.toBase58(),
            },
        });
        return response;
    }
    async getClaimTransactions(wallet, tokenMint) {
        const response = await this.bagsApiClient.post('/token-launch/claim-txs/v3', {
            feeClaimer: wallet.toBase58(),
            tokenMint: tokenMint.toBase58(),
        });
        const deserializedTransactions = response.map((tx) => {
            const decodedTransaction = bs58_1.default.decode(tx.tx);
            return web3_js_1.Transaction.from(decodedTransaction);
        });
        return deserializedTransactions;
    }
}
exports.FeesService = FeesService;
