// @bagsfm/bags-sdk/dist/services/incorporation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncorporationService = void 0;
const web3_js_1 = require("@solana/web3.js");
const bs58_1 = __importDefault(require("bs58"));
const base_1 = require("./base");
const validations_1 = require("../utils/validations");
class IncorporationService extends base_1.BaseService {
    constructor(apiKey, connection, commitment = 'processed') {
        super(apiKey, connection, commitment);
    }
    async startPayment(params) {
        const normalized = (0, validations_1.validateAndNormalizeStartPaymentParams)(params);
        const response = await this.bagsApiClient.post('/incorporate/start-payment', normalized);
        const decodedTransaction = bs58_1.default.decode(response.transaction);
        const transaction = web3_js_1.VersionedTransaction.deserialize(decodedTransaction);
        return {
            orderUUID: response.orderUUID,
            recipientWallet: response.recipientWallet,
            priceUSDC: response.priceUSDC,
            transaction,
            lastValidBlockHeight: response.lastValidBlockHeight,
        };
    }
    async incorporate(params) {
        const normalized = (0, validations_1.validateAndNormalizeIncorporateParams)(params);
        const { incorporationShareBasisPoint, ...rest } = normalized;
        const response = await this.bagsApiClient.post('/incorporate/incorporate', {
            ...rest,
            bedrockShareBasisPoint: incorporationShareBasisPoint,
        });
        return {
            tokenAddress: response.tokenAddress,
            incorporationStatus: response.incorporationStatus,
            founders: response.founders.map((f) => ({
                founderId: f.founderId,
                firstName: f.firstName,
                lastName: f.lastName,
                kycUrl: f.kycUrl,
                kycStatus: f.kycStatus,
                shareBasisPoint: f.shareBasisPoint,
                formUrl: f.bedrockFormUrl,
                pep: f.pep,
                ipAttributionAcknowledgedAt: f.ipAttributionAcknowledgedAt,
            })),
            incorporationShareBasisPoint: response.bedrockShareBasisPoint,
            category: response.category,
            twitterHandle: response.twitterHandle,
            preferredCompanyNames: response.preferredCompanyNames,
        };
    }
    async startIncorporation(params) {
        return this.bagsApiClient.post(`/incorporate/start-incorporation/${params.tokenAddress.toBase58()}`, {});
    }
    async list() {
        const response = await this.bagsApiClient.get('/incorporate/list');
        return response.map((project) => ({
            tokenAddress: project.tokenAddress,
            incorporationStatus: project.incorporationStatus,
            founders: project.founders,
            incorporationShareBasisPoint: project.bedrockShareBasisPoint,
            category: project.category,
            twitterHandle: project.twitterHandle,
            createdAt: project.createdAt,
            preferredCompanyNames: project.preferredCompanyNames,
            isReadyForIncorporation: project.isReadyForIncorporation,
        }));
    }
    async getDetails(params) {
        const response = await this.bagsApiClient.get(`/incorporate/details/${params.tokenAddress.toBase58()}`);
        return {
            tokenAddress: response.tokenAddress,
            incorporationStatus: response.incorporationStatus,
            incorporationShareBasisPoint: response.bedrockShareBasisPoint,
            category: response.category,
            twitterHandle: response.twitterHandle,
            createdAt: response.createdAt,
            preferredCompanyNames: response.preferredCompanyNames,
            isReadyForIncorporation: response.isReadyForIncorporation,
        };
    }
}
exports.IncorporationService = IncorporationService;
